Issue-1104_DocumentAndTestKeyConstructorsForCopyBehavior updated java…

DonResnik opened the pull request:
…doc and added unit tests for copy behavior
 Issue link:
https://github.com/apache/accumulo/issues/1104

keith-turner commented on core/src/main/java/org/apache/accumulo/core/data/Key.java:
```diff
@@ -119,7 +119,10 @@ public Key() {
    * Creates a key with the specified row, empty column family, empty column qualifier, empty column
-   * visibility, timestamp {@link Long#MAX_VALUE}, and delete marker false.
+   * visibility, timestamp {@link Long#MAX_VALUE}, and delete marker false.  This constructor creates
+   * a copy of row. If you don't want to create a copy of row, you should call
```
Could be made a bit shorter w/o loss of information.  Something like :

```
To avoid copying, use {@link Key#Key(byte[] row, byte[] cf, byte[] cq, byte[] cv, long ts, boolean deleted, boolean copy)}.
```

DonResnik replied:
code updated to use more concise comment

keith-turner commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +79,144 @@ public void testCopyData() {
+    assertNotSame(cq, kRowcolFamColQual.getColQualifier());
+    assertNotSame(cv, kRowcolFamColQual.getColVisibility());
+
+    assertNotSame(row, kRowcolFamColQual.getRowData().getBackingArray());
```
This pattern seems to be repeated, could create a function with these statements like :

```java
private assertNotSame(byte[] row, byte[] f, byte[] q, byte[] cv, Key k)
```

DonResnik replied:
great idea, this removed many lines of code

keith-turner commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +79,144 @@ public void testCopyData() {
+    
+    // test row, column family, column qualifier, column visibility constructor
+    assertNotSame(row, kRowcolFamColQualColVis.getRowBytes());
+    assertNotSame(cf, kRowcolFamColQualColVis.getColFamily());
```
Not sure, but I think different types are being compared (byte[] and Text)?  If so seems like they could never be the same.

DonResnik replied:
the Text objects are covered to byte[] before getting stored...if I am reading the code correctly. II added an equals method for the byte[] to make sure

keith-turner commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +79,144 @@ public void testCopyData() {
+    assertNotSame(cv, kRow.getColumnVisibilityData().getBackingArray());
+    
+    // test row, column family constructor
+    assertNotSame(row, kRowcolFam.getRowBytes());
```
Could make a method for this pattern.

DonResnik replied:
method was added

keith-turner commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +79,144 @@ public void testCopyData() {
+    assertNotSame(cq, kRowcolFam.getColQualifier());
+    assertNotSame(cv, kRowcolFam.getColVisibility());
+
+    assertNotSame(row, kRowcolFam.getRowData().getBackingArray());
```
Different types again?

ctubbsii commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +79,144 @@ public void testCopyData() {
+    Key kRowcolFamColQualColVisTimeStamp = new Key(row, cf, cq, cv, 5L);
+
+    // test row constructor
+    assertNotSame(row, kRow.getRowBytes());
```
Since these can be not the same for a variety of reasons, we probably want to check they are both not the same, and yet still equal.

DonResnik replied:
I added an equals method for the byte[] to show they are equal even when copied.

ctubbsii commented:
@DonResnik Thanks for working on this! Some hints on best practices for formatting git log messages can be found https://chris.beams.io/posts/git-commit/

For example, your commit message could be:

```
Fix #1105 Add tests and update docs for Key constructors

<more details here, if necessary>
```

DonResnik commented:
Thanks for the reviews and comments.  Please let me know if you see anything else that needs to be addressed.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/data/Key.java:
```diff
@@ -119,7 +119,11 @@ public Key() {
+   * visibility, timestamp {@link Long#MAX_VALUE}, and delete marker false.  This constructor creates
+   * a copy of row. 
+   * 
+   * To avoid copying, use {@link Key#Key(byte[] row, byte[] cf, byte[] cq, byte[] cv, 
```
Do you want this to be another paragraph?  Is so will need to add `<p>` to beginning.

DonResnik replied:
<p>added, thanks

keith-turner commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +80,96 @@ public void testCopyData() {
+    assertNotSameByteArray(kRowcolFamColQualColVisTimeStamp, row, cf, cq, cv, 5L);
+  }
+  
+  private void assertNotSameByteArray(Key key, byte[] row, byte[] cf, byte[]cq, byte[]cv, long timestamp) {
```
nothing is done with timestamp.

DonResnik replied:
removed timestamp from method signature

keith-turner commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +80,96 @@ public void testCopyData() {
+	  }
+	  if (key.getColFamily().length != 0) {
+		    assertNotSame(cf, key.getColFamily());
+		    assertNotSame(row, key.getColumnFamilyData().getBackingArray());
```
this should be `assertNotSame(cf`

DonResnik replied:
thanks, fixed

keith-turner commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +80,96 @@ public void testCopyData() {
+	  }
+	  if (key.getColQualifier().length != 0) {
+		    assertNotSame(cq, key.getColQualifier());
+		    assertNotSame(row, key.getColumnQualifierData().getBackingArray());
```
should use cq and not row

DonResnik replied:
thanks, fixed

keith-turner commented on core/src/test/java/org/apache/accumulo/core/data/KeyTest.java:
```diff
@@ -79,6 +80,96 @@ public void testCopyData() {
+    Key kRowcolFamColQualColVisTimeStamp = new Key(rowText, cfText, cqText, cvText, 5L);
+
+    // test row constructor
+    assertNotSameByteArray(kRow, row, cf, cq, cv, 5L);
```
The byte arrays passed here are not used to construct the key.

DonResnik replied:
created empty array to pass to test method when not all values are used

keith-turner replied:
I don't think this test with Text accomplishes anything.  The following statement will create an internal byte array in Text.

```java
Text rowText = new Text("r");
```

The byte array in `rowText` is always a different array instance than what the following statement creates.

```java
 byte[] row = "r".getBytes();
```

Therefore the byte array in `rowText` and the byte array `row` will never be the same.  But the reason they will never be the same has nothing to do with Key.

DonResnik replied:
I changed to goal of the Text test to ensure it creates the correct byte array in the constructor.

DonResnik commented:
I am trying to fix the travis error (trailing whitespace).  I added the Accumulo formatter to my local Eclipse and I added some other Eclipse options to remove trailing whitespace, but that did not change any files.  Is there any way to run the travis checks locally so I can make sure the formatting is good without pushing an update? Any ideas would be appreciated.  Thanks.

ctubbsii commented:
You should be able to run the command that Travis indicates is the failing one:
`mvn clean verify javadoc:jar -DskipITs`

DonResnik commented:
I ran the following

mvn clean verify javadoc:jar -DskipITs

It ran long enough to update the formatting (still waiting to see if it passes the Travis tests), but I got this error when trying to compile accumulo-core...can I fix this?  I did not make any changes to the Key class.  

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project accumulo-core: Compilation failure: Compilation failure: 
[ERROR] /Users/VandelayMacBookAir/git/accumulo/core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:[352,62] incompatible types: ? super java.util.Iterator<java.util.Map.Entry<org.apache.accumulo.core.data.Key,org.apache.accumulo.core.data.Value>> cannot be converted to java.util.Iterator<java.util.Map.Entry<org.apache.accumulo.core.data.Key,org.apache.accumulo.core.data.Value>>
[ERROR] /Users/VandelayMacBookAir/git/accumulo/core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:[352,35] incompatible types: bad return type in lambda expression
[ERROR]     cannot infer type-variable(s) F,T
[ERROR]       (argument mismatch; incompatible parameter types in lambda expression)

DonResnik commented:
FWIW, I got past all the build/format issues.

EdColeman commented:
If I understand the comments by @keith-turner it looks like the comment on testing byte[] has been corrected.  Not part of this pull request, but noticed that the test testVisibilityGetters is missing the @Test annotation - so it never runs.  It looks like a valid test and passes if I add @Test and run locally.

DonResnik commented:
Updated unit test per Ed Coleman review (thanks Ed)

ctubbsii commented:
I'm working on rebase-ing this to merge.

ctubbsii commented:
Thanks for the PR @DonResnik . Sorry it took so long to get this merged. Also, I know this isn't your first contribution, but I noticed you weren't listed on our people page as a contributor. If you wish to be added as a contributor to https://accumulo.apache.org/people/ , please open a pull request to add yourself at https://github.com/apache/accumulo-website/edit/master/pages/people.md and leave a reference to `apache/accumulo#1105` in your commit log.

If you intend to be a regular contributor to Accumulo projects, please consider subscribing to our developer mailing list (https://accumulo.apache.org/contact-us/) and introducing yourself. :smiley_cat:
